pkg/encryptedsecret/fetch_secret.go

```diff
@@ -0,0 +1,121 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+// +build !windows
+
+package encryptedsecret
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+)
+
+const payloadVersion = "1.0"
+
+type limitBuffer struct {
+	max int
+	buf *bytes.Buffer
+}
+
+func (b *limitBuffer) Write(p []byte) (n int, err error) {
+	if len(p)+b.buf.Len() > b.max {
+		return 0, fmt.Errorf("command output was too long: exceeded %d bytes", b.max)
+	}
+	return b.buf.Write(p)
+}
+
+func execCommand(inputPayload string) ([]byte, error) {
+	command := config.Datadog.GetString("secret_backend_command")
+	args := config.Datadog.GetStringSlice("secret_backend_arguments")
+
+	if command == "" {
+		return nil, fmt.Errorf("no secret_backend_command set: could not decrypt secret")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(),
+		time.Duration(config.Datadog.GetInt("secret_backend_timeout"))*time.Second)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, command, args...)
+	if err := checkRights(cmd.Path); err != nil {
+		return nil, err
+	}
+
+	cmd.Stdin = strings.NewReader(inputPayload)
+	// setting an empty env in case some secrets were set using the ENV (ex: API_KEY)
+	cmd.Env = []string{}
+
+	out := limitBuffer{
+		buf: &bytes.Buffer{},
+		max: config.Datadog.GetInt("secret_backend_output_max_size"),
+	}
+	cmd.Stdout = &out
+
+	err := cmd.Run()
+	if err != nil {
+		if ctx.Err() == context.DeadlineExceeded {
+			return nil, fmt.Errorf("error while running '%s': command timeout", command)
+		}
+		return nil, fmt.Errorf("error while running '%s': %s", command, err)
+	}
+	return out.buf.Bytes(), nil
+}
+
+type secret struct {
+	Value    string
+	ErrorMsg string `json:"error"`
+}
+
+// for testing purpose
+var runCommand = execCommand
+
+// fetchSecret receives a list of secrets name to fetch, exec a custom executable
+// to fetch the actual secrets and returns them.
+func fetchSecret(secrets []string) (map[string]string, error) {
+	payload := map[string]interface{}{
+		"version": payloadVersion,
+		"secrets": secrets,
+	}
+	jsonPayload, err := json.Marshal(payload)
+	if err != nil {
+		return nil, fmt.Errorf("could not serialize secrets to fetch password: %s", err)
+	}
+	output, err := runCommand(string(jsonPayload))
+	if err != nil {
+		return nil, err
+	}
+
+	passwords := map[string]secret{}
```

Comment:
call this `secrets` instead?

Reply:
+1 for `secrets` everywhere, as e.g. in the similar [K8s documentation](https://kubernetes.io/docs/concepts/configuration/secret/)